### PR TITLE
Migration to Nuxt 3

### LINKED ISSUE #411: Nuxt 3 migration
| Start Date | ETA        | Project Lead | Actual Ship Date |
| ---------- | ---------- | ------------ | ---------------- |
| 2023-06-05        | TBD        | @obulat           | TBD              |

## Description

Vue 2 is reaching end-of-life at the end of 2023[^1]. We need to update to Nuxt 3 to use Vue 3.

Node 16, which is required _to build_ the frontend app (Nuxt 2 applications can theoretically _run_ on Node >16) reaches end-of-life on September 11th, 2023.

## Documents

- [x] #2311
- [x] #2324

## Issues

- [x] PR in Nuxt 2.16.xx: https://github.com/nuxt/nuxt/pull/19518
- [x] #815 
- [x] #2232
- [x] #2234

## Prior Art

Several members of the team have attempted the migration previously but without success. This is one of such attempts: https://github.com/WordPress/openverse-frontend/pull/313
The Nuxt team has said that working on simplifying the migration is one of their priorities for 2023, so hopefully, it will be easier this time.

[^1]: Vue 2 LTS: https://v2.vuejs.org/lts/

### PR DESCRIPTION
**Pull request by obulat**
## Fixes

Fixes #411 by @obulat

## Description

This PR migrates the frontend app from Nuxt 2 to Nuxt 3. The following functionality works well:
- search
- single result pages
- related items
- collection pages
- analytics
- Playwright tests (visual regression and e2e)
- unit tests
- auth

### Fetching changes

Fetching media on `search`, `single-detail` and `collection` pages is done in `useAsyncFetch`. Previously, we were splitting the fetching between middleware for the server and the `useFetch` for the client fetching, to make sure that the server finishes fetching before sending the page to the client, and there are no client-server mismatches. While working on this PR, I realized that Nuxt requires some data (`payload`) returned from the `useAsyncFetch` function to cancel re-fetching of the same data on the client when hydrating (which was actually causing the mismatch).
We cannot call `showError` from inside the `useAsyncFetch`, so there's an error watcher that redirects to the error page after the `useAsyncFetch` in the components.

The search results are now fetched in the search layout. The reason for moving it from the child `search.vue` page was the fact that the server-rendered page would not update the result count in the header after the server fetches the media. This was causing the server-client mismatch.

### Global audio player
The global audio component is moved to the `app.vue`, the base component for all pages. The global audio would either keep playing on navigation, or stop playing when navigating from one single result page to the other. Previously, we were getting the active audio properties from media store or from the related media store. If the item wasn't found in the store, it would be set to null, and the player would be closed until the new item is fetched from the API. To fix this, I added the `detail` property to the `activeAudio` store, so that if there is an active audio, we always have its properties and can show it in the global player.

### API proxying
The API calls are now proxied through the Nuxt server route (`/api/...`). This allows us to send the auth credentials with _all_ of the Nuxt requests.

### TODO:
- [x] add tests for the API token flow [#d5716a4](https://github.com/WordPress/openverse/pull/3646/commits/d5716a44ce8b03f3b42ae12f90069b8dbb27da65)
- [x] add tests for fetching media [#9aa194c](https://github.com/WordPress/openverse/pull/3646/commits/9aa194c8503898edacae327c3ddbbf42940cdc5b)
- [ ] #3664 add Storybook
- [ ] #3694  fix the Playwright Docker build on a Mac (currently, it does not build because a rollup platform-dependent dependency is not installed)
- [ ] #3665 re-add Prometheus
- [ ] check that Sentry works correctly
- [ ] Ensure that the `.env` template file has all necessary variables
- [ ] Check that the Docker build sets the `.env` variables correctly.
- [ ] Add a note to increase the memory locally if there's an error.

## Testing Instructions

Go to https://staging.openverse.org and check that the deployed version works correctly. Read through the code and see that it makes sense.

### Locally

1. Remove the node_modules folder and re-install packages:
```
rm -rf frontend/node_modules && just node-install
```
2. Run the project with the following modified option:
```
export NODE_OPTIONS=--max-old-space-size=8192 && just frontend/run dev
```

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by AetherUnbound — APPROVED** [review]
I've looked through about 1/4th of the files, and the app runs locally for me! I have a few questions, but overall **great work** Olga!!

File: .github/actions/build-docs/action.yml
Comment (dhruvkb) [reply .github/actions/build-docs/action.yml:20]: #3664 is open to re-add Storybook so I suppose this is coming back. Also applies to similar steps related to Storybook in the CI + CD workflow.

File: frontend/.env.template
Comment (dhruvkb) [inline frontend/.env.template:10]: Safe to expose/use publicly as the default?

File: frontend/src/components/VAudioThumbnail/VAudioThumbnail.vue
Comment (dhruvkb) [inline frontend/src/components/VAudioThumbnail/VAudioThumbnail.vue:39]: This syntax is new.

File: frontend/src/components/VCheckbox/VCheckbox.vue
Comment (dhruvkb) [inline frontend/src/components/VCheckbox/VCheckbox.vue:71]: Having to replace logical boolean with a string is unfortunate. This however wouldn't work because setting `disabled=""` still disables the control.

> To represent a false value, the attribute has to be omitted altogether.
> &mdash; [The spec](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attribute)

File: frontend/src/components/VCollectionHeader/meta/VCollectionHeader.stories.mdx
Comment (dhruvkb) [inline frontend/src/components/VCollectionHeader/meta/VCollectionHeader.stories.mdx:11]: Should have a lint for this. Thanks for catching!

**Review by dhruvkb — APPROVED** [review]
The changes make sense to me. I have left some comments but overall, I am excited to unblock this!

**Review by AetherUnbound — CHANGES_REQUESTED** [review]
Noting here that we may want to confirm/resolve the memory leak on staging before we merge this (just so we don't do it accidentally!).

File: frontend/src/components/VAudioThumbnail/VAudioThumbnail.vue
Comment (obulat) [reply frontend/src/components/VAudioThumbnail/VAudioThumbnail.vue:39]: Do you mean the `#imports`, @dhruvkb ?

File: frontend/src/components/VCollectionHeader/meta/VCollectionHeader.stories.mdx
Comment (obulat) [reply frontend/src/components/VCollectionHeader/meta/VCollectionHeader.stories.mdx:11]: We probably don't have it set up for mdx files. I think this story should also have been deleted, though.

File: frontend/test/unit/specs/utils/get-additional-sources.spec.ts
Comment (krysal) [inline frontend/test/unit/specs/utils/get-additional-sources.spec.ts:1]: Good thing this small change is all that's needed for some tests!

File: frontend/src/components/VAudioThumbnail/VAudioThumbnail.vue
Comment (krysal) [reply frontend/src/components/VAudioThumbnail/VAudioThumbnail.vue:39]: This is the first time I've seen that, too! 😮

File: packages/eslint-plugin/package.json
Comment (obulat) [reply packages/eslint-plugin/package.json:26]: They should, great catch!

File: frontend/tsconfig.json
Comment (obulat) [reply frontend/tsconfig.json:1]: Nuxt creates a `tsconfig.json` file that, in addition to setting TypeScript properties, also adds all of the paths of the auto-imported modules. Even though we are disabling auto-import, we do use the paths from this file. Nuxt documentation tells to create the `tsconfig.json` that extends this auto-generated file.

**Comment by sarayourfriend** [comment]
@obulat Can we close this PR in the meantime, until we get ready to attempt this again (or do something else entirely)?

I'll close it proactively, but please reopen it with a note why we need to keep it drafted long-term, if that is the case.

**Pull request closed without merging**
